packages/components/src/higher-order/with-filters/index.js

```diff
@@ -24,45 +24,82 @@ const ANIMATION_FRAME_PERIOD = 16;
  */
 export default function withFilters( hookName ) {
 	return createHigherOrderComponent( ( OriginalComponent ) => {
-		return class FilteredComponent extends Component {
-			/** @inheritdoc */
+		const namespace = 'core/with-filters/' + hookName;
+
+		/**
+		 * Since filtering is applied to the component, each filtered instance
+		 * can reuse a shared reference to the definition. This optimizes to
+		 * avoid excessive calls to `applyFilters` when many instances exist.
+		 *
+		 * @type {Component}
+		 */
+		let FilteredComponent = applyFilters( hookName, OriginalComponent );
+
+		class FilteredComponentRenderer extends Component {
 			constructor( props ) {
 				super( props );
 
-				this.onHooksUpdated = this.onHooksUpdated.bind( this );
-				this.Component = applyFilters( hookName, OriginalComponent );
-				this.namespace = uniqueId( 'core/with-filters/component-' );
-				this.throttledForceUpdate = debounce( () => {
-					this.Component = applyFilters( hookName, OriginalComponent );
```

Comment:
> @aduth, I saw this question in my inbox. I can't see it anymore in the original PR. I just wanted to confirm that it is triggered here and in your optimized version properly.

Yes, I deleted it once I noticed I'd overlooked the re-assignment in the debounce callback.

> The benefit of calling `applyFilters` inside of the `throttledForceUpdate` method is that it will be executed only once when multiple calls of add/remove filters occurs in the same small period of time.

Yeah, I can see that being a good thing. It might require that we do the refactoring I'd pondered at https://github.com/WordPress/gutenberg/pull/13231#discussion_r245809765 ?